rescan.go

```diff
@@ -482,24 +444,28 @@ func rescan(chain ChainSource, options ...RescanOption) error {
 	//
 	// TODO(wilmer): refactor this and handleBlockDisconnected into their
 	// own methods.
-	handleBlockConnected := func(ntfn *blockntfns.Connected) error {
+	handleBlockConnected := func(ntfn *blockntfns.Connected, isRetry bool) error {
+		// If we have any blocks to retry and the notification doesn't
+		// correspond to the current block being retried, we'll queue it
+		// up to retry after.
+		newHeader := ntfn.Header()
+		if len(retryBlocks) > 0 {
+			retryHeader := retryBlocks[0].Header()
+			if retryHeader.BlockHash() != newHeader.BlockHash() {
+				log.Debug("Stashing", ntfn)
+				retryBlocks = append(retryBlocks, ntfn)
+				return nil
+			}
+		}
+
 		// If we've somehow missed a header in the range, then we'll
 		// mark ourselves as not current so we can walk down the chain
 		// and notify the callers of blocks we may have missed.
-		//
-		// It's possible due to the nature of the current subscription
-		// system that we get a duplicate block. We'll catch this and
-		// continue forwards to avoid an unnecessary state transition
-		// back to the !current state.
-		header := ntfn.Header()
-		if header.PrevBlock != curStamp.Hash &&
-			header.BlockHash() != curStamp.Hash {
-
+		if newHeader.PrevBlock != curStamp.Hash {
```

Comment:
It would have been nice to have refactoring that doesn't change behavior in a separate commit, to get a better overview on the fix itself.

Reply:
Done.